src/compute-client/src/protocol/response.rs

```diff
@@ -102,11 +102,11 @@ pub enum ComputeResponse<T = mz_repr::Timestamp> {
     ///
     /// For each subscribe that was installed by a previous [`CreateDataflows` command], the
     /// replica must emit [`Batch`] responses that cover the entire time interval from the
-    /// subscribe dataflow's `as_of` until the subscribe advances to the empty frontier or is
+    /// minimum time until the subscribe advances to the empty frontier or is
     /// dropped. The time intervals of consecutive [`Batch`]es must be increasing, contiguous,
     /// non-overlapping, and non-empty. All updates transmitted in a batch must have times within
-    /// that batch’s time interval. The `upper` of the first [`Batch`] of a subscribe must not be
-    /// less than that subscribe's initial `as_of` frontier.
+    /// that batch’s time interval. The `upper` of the first [`Batch`] of a subscribe must
+    /// exactly the subscribe's initial `as_of` frontier, and it must be empty.
```

Comment:
@frankmcsherry @teskje @antiguru this deserves extreme scrutiny. This was the observed behavior of `SUBSCRIBE` in my several tests, but I didn't go digging to verify that this was 100% guaranteed by the current sink operator.

Separately, I can imagine that compute doesn't want to provide this guarantee, and instead wants adapter to provide it. That doesn't seem crazy but .. doing it that way requires adding untestable code to adapter, since I couldn't actually figure out how to get compute to return subscribe output that didn't start with one empty batch from [0, as_of).

Reply:
I'll defer to @teskje and @antiguru though. It wouldn't be the first untested code in MZ. :D But perhaps the path forward is to do some additional work that drops that first message, updates the guarantees, and any points that rely on the assumption.

Would it make sense to add an assert in Adapter that expects the first batch to be `[0, ASOF)`, so that a violation of this contract is immediately caught?

Reply:
Looking at the code, I don't see anything that would obviously guarantee that a) the first batch as the `as_of` as its `upper` or b) the first batch is empty. The only things I would be comfortable to expect from the current code are a) the first batch's `lower` is the minimum time and b) the first batch's `upper` is >= the `as_of`.

I'm not saying the current code doesn't guarantee the behavior described above, just that it's not obvious. If we want, it should be pretty easy to add a couple lines to `SubscribeProtocol` to make it obvious. Specifically, we could make `SubscribeProtocol` immediately send that first empty batch upon instantiation.

I agree with Frank that such an initial message would be redundant. The only thing it does is tell you "this subscribe is not going to produce any batches before `as_of` anymore", which is something you already know when creating a subscribe with that `as_of`. On the other hand, the subscribe protocol already contains a couple warts (see https://github.com/MaterializeInc/materialize/issues/16274, https://github.com/MaterializeInc/materialize/issues/17781), so it's probably fine to add this one as well, as long as we clean up eventually.